pkg/prober/prober.go

```diff
@@ -0,0 +1,111 @@
+package prober
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/common/route"
+)
+
+const (
+	healthyEndpointPath  = "/-/healthy"
+	readyEndpointPath    = "/-/ready"
+	okProbeText          = "thanos %s is %s"
+	probeErrorHTTPStatus = 500
+	initialErrorText     = "thanos %s is initializing"
+)
+
+// Prober represents health and readriness status of given compoent.
+type Prober struct {
+	componentMtx sync.Mutex
+	component    string
+	readyMtx     sync.Mutex
+	readiness    error
+	healthyMtx   sync.Mutex
+	healthiness  error
+}
+
+// SetComponent sets component name of the Prober displayed in responses.
+func (p *Prober) SetComponent(component string) {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	p.component = component
+}
+
+// NewProber returns Prober reprezenting readiness and healthiness of given component.
+func NewProber(component string) *Prober {
+	initialErr := fmt.Errorf(initialErrorText, component)
+	prober := &Prober{}
+	prober.SetComponent(component)
+	prober.SetNotHealthy(initialErr)
+	prober.SetNotReady(initialErr)
+	return prober
+}
+
+// NewProbeInRouter returns new Prober which registers it's ready and health endpoints to given router.
+func NewProbeInRouter(component string, router *route.Router) *Prober {
+	prober := NewProber(component)
+	router.Get(healthyEndpointPath, prober.probeHandlerFunc(prober.IsHealthy, "healthy"))
+	router.Get(readyEndpointPath, prober.probeHandlerFunc(prober.IsReady, "ready"))
+	return prober
+}
+
+// NewProbeInMux returns new Prober which registers it's ready and health endpoints to given mux.
+func NewProbeInMux(component string, mux *http.ServeMux) *Prober {
+	prober := NewProber(component)
+	mux.HandleFunc(healthyEndpointPath, prober.probeHandlerFunc(prober.IsHealthy, "healthy"))
+	mux.HandleFunc(readyEndpointPath, prober.probeHandlerFunc(prober.IsReady, "ready"))
+	return prober
+}
+
+func (p *Prober) probeHandlerFunc(probeFunc func() error, probeType string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		err := probeFunc()
+		if err == nil {
+			if _, e := io.WriteString(w, fmt.Sprintf(okProbeText, p.component, probeType)); e == nil {
+			}
```

Comment:
I'm not sure how to handle this.. Should I propagate a logger in here and log the error? 
I have nowhere to return it here :/

Reply:
Eventually I added the logging and also debug logs for state changes so it can be easier debugged possibly.